go/pkg/vpnkit/vmnet.go

```diff
@@ -0,0 +1,588 @@
+func defaultInitMessage() *InitMessage {
+	magic := [5]byte{'V', 'M', 'N', '3', 'T'}
+	version := uint32(22)
+	var commit [40]byte
+	copy(commit[:], []byte("0123456789012345678901234567890123456789"))
+	return &InitMessage{magic, version, commit}
+}
+
+// Write marshals an init message to a connection
+func (m *InitMessage) Write(c net.Conn) error {
+	if err := binary.Write(c, binary.LittleEndian, m.magic); err != nil {
+		return err
+	}
+	if err := binary.Write(c, binary.LittleEndian, m.version); err != nil {
+		return err
+	}
+	if err := binary.Write(c, binary.LittleEndian, m.commit); err != nil {
+		return err
+	}
+	return nil
+}
+
+// readInitMessage unmarshals an init message from a connection
+func (v *Vmnet) readInitMessage() (*InitMessage, error) {
+	m := defaultInitMessage()
+	if err := binary.Read(v.conn, binary.LittleEndian, &m.magic); err != nil {
+		return nil, err
+	}
+	if err := binary.Read(v.conn, binary.LittleEndian, &m.version); err != nil {
+		return nil, err
+	}
+	if err := binary.Read(v.conn, binary.LittleEndian, &m.commit); err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (v *Vmnet) negotiate() error {
+	m := defaultInitMessage()
+	if err := m.Write(v.conn); err != nil {
+		return err
+	}
+	remoteVersion, err := v.readInitMessage()
+	if err != nil {
+		return err
+	}
+	v.remoteVersion = remoteVersion
+	return nil
+}
+
+// Ethernet requests the creation of a network connection with a given
+// uuid
+type Ethernet struct {
+	uuid uuid.UUID
+}
+
+// Write marshals an Ethernet message
+func (m *Ethernet) Write(c net.Conn) error {
+	ty := uint8(1)
+	if err := binary.Write(c, binary.LittleEndian, ty); err != nil {
+		return err
+	}
+	u, err := m.uuid.MarshalText()
+	if err != nil {
+		return err
+	}
+	if err := binary.Write(c, binary.LittleEndian, u); err != nil {
+		return err
+	}
+	padding := uint32(0)
+	if err := binary.Write(c, binary.LittleEndian, padding); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Vif represents an Ethernet device
+type Vif struct {
+	MTU           uint16
+	MaxPacketSize uint16
+	ClientMAC     net.HardwareAddr
+	IP            net.IP
+	conn          net.Conn
+}
+
+func (v *Vmnet) readVif() (*Vif, error) {
+	var MTU, MaxPacketSize uint16
+
+	if err := binary.Read(v.conn, binary.LittleEndian, &MTU); err != nil {
+		return nil, err
+	}
+	if err := binary.Read(v.conn, binary.LittleEndian, &MaxPacketSize); err != nil {
+		return nil, err
+	}
+	var mac [6]byte
+	if err := binary.Read(v.conn, binary.LittleEndian, &mac); err != nil {
+		return nil, err
+	}
+	padding := make([]byte, 1+256-6-2-2)
+	if err := binary.Read(v.conn, binary.LittleEndian, &padding); err != nil {
+		return nil, err
+	}
+	ClientMAC := mac[:]
+	conn := v.conn
+	var IP net.IP
+	return &Vif{MTU, MaxPacketSize, ClientMAC, IP, conn}, nil
+}
+
+// ConnectVif returns a connected network interface with the given uuid
+func (v *Vmnet) ConnectVif(uuid uuid.UUID) (*Vif, error) {
+	e := Ethernet{uuid}
```

Comment:
Would it make sense to support `Preferred_ipv4` here too? The IP should be known, but the command could fail from server with IP in use or invalid IP.